### DISCUSSION TITLE
Frustrated about no auto-save and backup

### DISCUSSION THREAD

**Discussion by Cosmorider** [excalidraw/excalidraw#6463] [General]
So, about 2 or 3 times I've worked for hours to create drawings and I then forget to save them.

Now, when using any Microsoft application, this wouldn't be an issue, they realize how important saving users their sanity and time is, so they always try to recover what the user has been working on.

However, for Excalidraw, there is no backup. There is no prompt that asks you if you want to save when exiting it. Sometimes when you return to the Excalidraw everything is somehow magically there, other times it's not and you hope to God that you've saved your work.

I'm just wondering, why would anybody want to use this application given these limitations? Do people enjoy suffering and misery in their life? Just curious.

I didn't find any other similar discussion other than a person losing 2 weeks of their work.

**Comment by sojusnik** [comment]
I feel with you! Experienced the same. This was brought up earlier, see f.i.

https://github.com/excalidraw/excalidraw/issues/2733
https://github.com/excalidraw/excalidraw/pull/3047

> Sometimes when you return to the Excalidraw everything is somehow magically there, other times it's not and you hope to God that you've saved your work.

Can confirm that this works very unreliable. Sometimes the cached version upon opening the browser or PWA is up-to-date, sometimes an older version of the canvas is loaded without changes made in the previous session. This is very frustrating indeed.

Let's hope that the devs will catch up soon!

**Reply by Cosmorider** [reply]
Many people that I respect that are accomplished in their field have used Excalidraw so I just assumed that it would be very reliable, but it seems that it is only reliable within the confines of something like Obsidian where it seems to be automatically saved. I definitely do not recommend anybody to use the actual stand-alone application.

**Reply by sojusnik** [reply]
I also was thinking about using Excalidraw in Obsidian only, because of this issue. And also, because you can search there, which is currently missing in vanilla Excalidraw, as mentioned [here](https://github.com/excalidraw/excalidraw/issues/6066).

**Comment by dwelle** [comment]
Hello. We're autosaving to the browser (localStorage), so what you're seeing should definitely not happen.

There are 2 cases where loss of data may occur:

1) you go over the 5MB localStorage limit. In that case, it should not save the data above that 5MB limit unless you pasted 5MB on a clean canvas in one go, in which case none of it would get saved.
2) you clear your browser cache/data (or the browser does it for you — I believe Safari may still be doing it after 7 days of inactivity or something like that).

Could you right-click on canvas, open "stats for nerds" and check the canvas size?

I agree we should show an error message if you go over 5MB, which we currently aren't doing.

**Reply by Cosmorider** [reply]
Yeah, I remember I was copying and pasting a lot of photos into Excalidraw, I then saved everything, and when I loaded it up all of the photos were gone.

I probably went over the storage limit, I was pasting a lot of photos from the snipping tool.

I didn't clear the browser cache/data, I use Brave.

Yeah it would be appreciated if that error message would show, I usually create huge canvases with a lot of detail.

**Reply by sojusnik** [reply]
In my case the storage limit is 2.5MB, as shown in the picture below and I don't clear the cache/data in the browser I'm using Excalidraw with (it's Brave as a PWA) which is associated with Excalidraw, but I'm sure that already several times my changes weren't saved in the local storage, so I accidentally continued to work on an earlier version of my canvas, when reopening my browser.

I have the assumption that this may happen when Excalidraw is updating itself, but can't reproduce that.

![grafik](https://user-images.githubusercontent.com/1245718/232329444-94343483-2928-4416-8c3c-cf9d2e03ed07.png)

My current workaround is to always save every change to a local file and open this file, when I start the Excalidraw PWA, but sometimes I forget to load the local file or I forget to save to the local file when closing, because you still aren't reminded to save the changes made to the local file upon exit, which is kinda frustrating.

**Reply by dwelle** [reply]
> I have the assumption that this may happen when Excalidraw is updating itself, but can't reproduce that.

This has happened to me in the past as well, but none in the team could reproduce. Warrants further investigation for sure.

Other possibility (since you both use Brave) is that Brave is doing something to storage since it's very aggressive in privacy, so it's possible it deletes site data more aggressively than others.

**Reply by sojusnik** [reply]
> > I have the assumption that this may happen when Excalidraw is updating itself, but can't reproduce that.
> 
> This has happened to me in the past as well, but none in the team could reproduce. Warrants further investigation for sure.

Good to know that I wasn't hallucinating!

> Other possibility (since you both use Brave) is that Brave is doing something to storage since it's very aggressive in privacy, so it's possible it deletes site data more aggressively than others.

Not likely, since it then would delete the whole canvas and not only the recent changes, I suppose.

Can we expect in the near future at least a reminder to save to the local file, when the canvas is edited through a local file, or an auto-save feature to the local file, when closing Excalidraw, to mitigate that aforementioned bug we weren't able to reproduce?

**Reply by dwelle** [reply]
> Can we expect in the near future at least a reminder to save to the local file, when the canvas is edited through a local file, or an auto-save feature to the local file, when closing Excalidraw, to mitigate that aforementioned bug we weren't able to reproduce?

We can do that. But on Brave the experience would be really subpar as they removed support for filesystem API. So you wouldn't be able to save back to the same file.

So unless we detect the filesystem API is supported, we'd not show the reminder.

In brave you can do in flags (brave://flags):

![image](https://user-images.githubusercontent.com/5153846/233011382-086d216f-f2a7-4662-a756-a87ad2c52abd.png)

**Reply by sojusnik** [reply]
> But on Brave the experience would be really subpar as they removed support for filesystem API. So you wouldn't be able to save back to the same file.

You mean, if the API isn't enabled manually? I'm using Brave with the "File System Access API" enabled for months now and don't have any problems opening a local file, editing it and then saving to the same file.

I also think that the autosave function should only be triggered when the whiteboard is opened through a local file, so when closing the whiteboard, all changes should be saved to the opened local file (not a new one) automatically, without asking. Maybe it would help also, if the autosave function would also save to the local file every f.i. 5 minutes and not only upon closing Excalidraw.
